networkx/algorithms/community/modularity_max.py

```diff
@@ -234,7 +218,97 @@ def greedy_modularity_communities(G, weight=None, resolution=1, n_communities=1)
+        yield list(communities.values())
+
+
+def greedy_modularity_communities(
+    G, weight=None, resolution=1, cutoff=None, best_n=None, **aliases
+):
+    """
+    Parameters
+    ----------
+    G : NetworkX graph
+
+    weight : string or None, optional (default=None)
+        The name of an edge attribute that holds the numerical value used
+        as a weight.  If None, then each edge has weight 1.
+        The degree is the sum of the edge weights adjacent to the node.
+
+    resolution : float, optional (default=1)
+        If resolution is less than 1, modularity favors larger communities.
+        Greater than 1 favors smaller communities.
+
+    cutoff : int or None, optional (default=None)
+        The number of communities in the partition at which the merging process
+        will stop even if a maximum is not yet reached.  If None, continue until
+        a maximum modularity is reached or all nodes are in a single community.
+
+    best_n : int or None, optional (default=None)
+        Force community merging to continue until `best_n` communities remain
+        even if the modularity decreases.
+        If ``None``, don't force it to continue beyond a maximum.
+
+    Returns
+    -------
+    partition: list
+        A list of frozensets of nodes, one for each community.
+        Sorted by length with largest communities first.
+    """
+    if cutoff is not None:
+        if (cutoff < 1) or (cutoff > G.number_of_nodes()):
+            raise ValueError(
+                f"cutoff must be between 1 and {G.number_of_nodes()}. Got {cutoff}"
+            )
+    else:
+        cutoff = 1
+    if best_n is not None:
+        if (best_n < 1) or (best_n > G.number_of_nodes()):
+            raise ValueError(
+                f"best_n must be between 1 and {G.number_of_nodes()}. Got {best_n}"
+            )
+    else:
+        best_n = G.number_of_nodes()
+    if "n_communities" in aliases:
+        import warnings
+
+        warnings.warn(
+            "kwarg ``n_communities`` in greedy_modularity_communities is deprecated"
+            "and will be removed in version 3.0.   Use ``cutoff`` instead.",
+            DeprecationWarning,
+        )
+        cutoff = aliases.pop("n_communities")
+
+    # retrieve generator object to construct output
+    community_gen = greedy_modularity_communities_generator(
+        G, weight=weight, resolution=resolution
+    )
+
+    # construct the first two best partitions
+    prv = next(community_gen)
+    if len(prv) == 1:
+        return prv
+    nxt = next(community_gen)
+
+    # calculate modularity for these partitions/communities
+    mod_prv = modularity(G, prv, resolution=resolution, weight=weight)
+    mod_nxt = modularity(G, nxt, resolution=resolution, weight=weight)
+    dq = mod_nxt - mod_prv
+
+    while len(prv) > 1:
+        # additional breaking critera based on kwargs
+        if (dq > 0 and len(prv) <= cutoff) or (dq < 0 and len(prv) <= best_n):
+            break
```

Comment:
I see what you are saying about wanting to return `prv` when `dq<0`.  That makes this more complicated.
I'm still having trouble seeing the logic all the way through -- so I know that if I look at this in 6 months I won't see it. :}

Why is it an advantage to split the conditions for stopping into line 299 instead of in the while statement?
I'm also having trouble with why `len(prv) <= cutoff` is the right inequality here?  Maybe I just need to come back later when my brain has reset. :}

Are each of the stopping conditions contained in a test to make sure we've got all cases covered? If so, maybe I just need to look at the tests.
Thanks!

Reply:
I agree it's a bit confusing. The main logic is that in the generator, you can break the while loop before merging two new communities as only the difference `dq` can be calculated. Here, we generate the communities in any case and then check later if the newly constructed communities `nxt` are better or worse. If better, we keep going, if worse we return the better community set `prv`. I can add a comment maybe, so this will become clearer?

Resp ln 299: You can add it in the while statement, but I thought this way is more comprehensive. Adding it to the while statement would include then in total 4 conditions. The first one `if True` and the remaining three `if not False`.
But, as you mentioned before, that's taste. I don't really care and can include all of them in the while statement.

True, it should be an equality, ie. `==,` not `<=`. That's somewhat the same because in each iteration we're just reducing `len(prv)` and initially it has to be `>=` (otherwise we'd end up with an error [here](https://github.com/networkx/networkx/pull/5227/files#diff-a9c150aa8f828e5c4485350f618c6a50626ab1d2463ef0ce84792ab06763bf97R258-R261)). Nevertheless, should be changed to `==`. Good catch!

Yup, all parameters should have tests.
The new kwargs have tests:
`cutoff`: test_cutoff_parameter
`n_best`: test_n_best()
Tests for the other parameters were not changed here (as we expect the same result with the new function). They are covered in the tests, too.

Reply:
Ack -- OK now I am understanding how the generator and the original functions differ. I think this method means we are recomputing the modularity at every step of the merge.  That un-does all the work that the original function does to compute and track `dq` instead of computing the modularity. The complexity of the resulting algorithm is much worse that the previous function.

We have to make use of the `dq` computed in the generator so we can avoid any modularity computation in the outer function. Does this make sense? 

I'm also trying to save the construction of `nxt` until we know whether it is better than `prv`. The old code did that by looking at `dq` for merging (u, v) before actually merging (u,v). 

I've been trying to find a way to get the generator to output `dq` and the communities. But that also requires constructing the new communities before the outer function knows if it needs them.  Here's a crazy suggestion:
We could have the generator yield `dq` values alternating with the partition `communities.values()`. It first yields the partition with each node in isolation. Then it yields the `dq` value of the next step. Then it yields the partition after that step. Then the `dq` for the next step, and it continues until it returns one community with all the nodes.  This means we have two yield statements inside the generator's `while len(H) > 1:` loop. One right after computing `dq = -negdq`. The other just after merging the `communities`. That way we never compute more in the generator than we need to.

The logic in the outer function seems to make more sense to me then too.  Something like:
```python
    stop_at_max = best_n is None
    if not stop_at_max:
        cutoff = best_n

    comm = next(communities_gen)
    while len(comm) > cutoff:   # stop computing when len(comm) == cutoff
        dq, u, v = next(communities_gen)
        if stop_at_max and dq < 0:    # stop at the max_mod, unless best_n gives a size to go to
            break
        comm = next(communities_gen)
    return sorted(comm, key=len, reverse=True)
```

Reply:
I understand your suggestion and it does make sense, but indeed somewhat crazy! I'm not sure if that makes using the generator outside the new function favorable, as the alternating outputs might be somewhat confusing. Would require careful re-writing the whole documentation in the beginning of the function. But I see your point. How expensive is the calculation of modularity? And how expensive is understanding altering outputs of a generator? Guess that's much harder to put a price on...

Anyways, I can try and play a bit with the code and come up with a draft later today.

Reply:
Unfortunately, calculating modularity is very costly -- I don't recall what the complexity is, but basically computing modularity requires all edges of the graph while dq requires just looking at the neighbors of u and v. It's Ok for small graphs, but increasing quickly for bigger graphs.

To make the generator user-friendly we might need to do something like make it private (add underscore to the front of its name and take it out of `__all__`) and then make a front-end generator that yields just the communities. Both the front-end generator and the greedy function would call the private generator.  A few paragraphs of docs explaining the private generator output (in the private generator's docstring will be needed for sure.

Reply:
Didn't construct a front-end generator yet. Let's first finish these two.